Technical implementation proposal for Sponsorship in CLI (#14997)
NuGet/Home: accepted/2026/implementing-sponsorship-in-cli.md

# Implementing NuGet.Client CLI Sponsorship Support

- Author: [@kalebfik](https://github.com/kalebfik)
- GitHub Issue: [NuGet/NuGetGallery#10703](https://github.com/NuGet/NuGetGallery/issues/10703)

## Summary

Add sponsorship reporting to the NuGet CLI for installed packages.

The command surface will be presented as `dotnet package list --sponsor`

It will reuse the same metadata-fetch and rendering pipeline, with:

- A `Sponsor` column in console output
- A `sponsorablePackages` array in JSON output

## Explanation

### Functional Explanation

When a user runs `dotnet package list --sponsor`, the CLI will examine the projects/solutions packages and retrieve sponsorship links from nuget.org. 
The results are then grouped by project and displayed in the console, and optionally in JSON format. 

> Packages that have multiple links are returned in the order received by nuget.org

Reviewer: nit
```suggestion
Packages that have multiple links are returned in the order received by nuget.org.
```

The current proposal will receive sponsorship information from nuget.org. 
The new `metadata` and `sponsorshipUrls` property (shown in technical explanation) do not automatically propagate in other sources, resulting in those sources not receiving sponsorship information.   

> Sponsorship information is currently queried from nuget.org

Reviewer: ```suggestion
Sponsorship information is currently queried from nuget.org.
```

> In debug builds only, the environment variable `NUGET_SPONSOR_ENDPOINT_OVERRIDE` can be configured to https://apidev.nugettest.org/v3/index.json (for DEV) and https://apiint.nugettest.org/v3/index.json (INT).

Reviewer: I'd move this down to the end of the Functional explanation. Let's say something about this being a testing strategy for development of the feature. Maybe a new subsection like `#### Testing strategy` or something.

The command resolves packages from NuGet configuration and --source. 

> Packages that do not contain sponsorship links or sources that don't support sponsorship data will be omitted from the report. 

Reviewer: Omitted is technically correct, but we want to show a message in CLI and `problems` in JSON output.

> The console will display:

Reviewer: Can the condition be written out here? 

eg,
```suggestion
When nuget.org is a configured package source or is specified with `--source`, but no installed packages provided sponsorship data, the console will display:
```

```text
There are no sponsorship details found at https://api.nuget.org/v3/index.json
```

When nuget.org is not configured as a source and the command is used, the console will display a message:

```text
For sponsorship details, configure nuget.org as a package source, or allow access to nuget.org by specifying `--source https://api.nuget.org/v3/index.json`
```

> The JSON output will include an empty `sponsorablePackages` array for those projects.

Reviewer: I think if nuget.org isn't configured, we should output an Error in the JSON.

I'd like to see CLI sample output and JSON sample output for all messages.
Something like:

```json
"problems": [
  {
    "level": "warning",
    "text": "There are no sponsorship details found at https://api.nuget.org/v3/index.json"
  }
]
```
```json
"problems": [
  {
    "level": "error",
    "text": "For sponsorship details, configure nuget.org as a package source, or allow access to nuget.org by specifying `--source https://api.nuget.org/v3/index.json`"
  }
]

The default console output will add a `Sponsor` column. 
The first URL will appear besides the package ID and each additional URL appears below the previous one. 

> When using `--format json`, those same packages are returned in JSON format, identified by `sponsorablePackages` and with links stored in each package's `urls` array. 

Reviewer: Move this json commentary below the CLI output.

`dotnet package list --sponsor` output: 

```text
Top-level Package        Sponsor
> Contoso.Tools          https://github.com/sponsors/username
> Contoso.Utility        https://github.com/sponsors/username
                         https://domain.com/sponsor
```

`dotnet package list --sponsor --format json` output: 

```json
{
  "version": 1,
  "parameters": "--sponsor",
  "sources": [
    "https://api.nuget.org/v3/index.json"
  ],
  "projects": [
    {
      "path": "/path/to/Contoso.App.csproj",
      "sponsorablePackages": [
        {
          "id": "Contoso.Forms",
          "urls": [
            "https://sponsordomain.com/sponsors/contoso",
          ]
        },
        {
          "id": "Contoso.Utility",
          "urls":[
            "https://github.com/sponsors/username",
            "https://domain.com/sponsor"
          ]
        }
      ]
    }
  ]
}
```

**Package Source Mapping and using `--source`**

The proposed implementation does not apply Package Source Mapping (PSM).

> When PSM is configured, the command will not make sponsorship metadata requests unless nuget.org is stated using `--source https://api.nuget.org/v3/index.json`.

Reviewer: We talked about just erroring in this case. Please add CLI/JSON samples.

```suggestion
When PSM is configured, the command will error unless nuget.org is stated using `--source https://api.nuget.org/v3/index.json`.
```

### Technical explanation

`dotnet package list --sponsor` extends the existing package-list report pipeline with a new report type. 

Planned updates:

- Add `--sponsor` handling in `ListPackageCommand`, `ListPackageArgs`, and `ReportType`, executed through `ListPackageCommandRunner`

> - Deserialize `sponsorshipUrls` through `RegistrationIndex`, and propagate them to `PackageSearchMetadataRegistration` in `PackageMetadataResourceV3`

Reviewer: Before deserializing, will you be validating the configured sources for nuget.org?

We now also decided to support a test environment envvar, so mentioning use of `EnvironmentVariableWrapper.Instance` to override the hard-coded nuget.org would be good.

- Preserve sponsorship data through `InstalledPackageReference` and `ListReportPackage`.
- Render console sponsorship output through `ListPackageConsoleRenderer` and `ProjectPackagesPrintUtility`.
- Render JSON sponsorship output through `ListPackageJsonRenderer`.

#### Data assumptions

- The client reads optional `sponsorshipUrls` as `IReadOnlyList<string>`.
- Only packages with one or more sponsorship urls are included. 
- URLs preserve the order that is returned by nuget.org in both the console and JSON.
- At this time, nuget.org enforces a maximum of 10 URLs shown per package.; client will not be validating URLs, it will only show what the server provides. 

**Example Registration Index:**

```json
{ 
  "@id": "https://api.nuget.org/v3/registration5-gz-semver2/contoso.webapi.client/index.json",
  "@type": [
    "catalog:CatalogRoot", 
    "PackageRegistration", 
    "catalog:Permalink"
  ],
  "commitId": "afa91af1-9505-41b8-ad75-eab8e613db14",
  "commitTimeStamp": "2026-04-10T00:15:25.1492389+00:00",
  "count": 2,
  "metadata":

>     // ** Start of proposal ** //

Reviewer: I know the server change isn't part of this spec, but I think we should highlight that "metadata" is a new node as well.

```suggestion
    // ** Start of proposal ** //
    "metadata":
```

    "sponsorshipUrls": [
      "https://github.com/sponsors/contoso"
      ]
    // ** End of proposal **//
}
```

#### Source transport dependency and client abstraction

This proposal assumes the CLI consumes sponsorship data from the Registration API.

> Missing or empty `sponsorshipUrls` use the empty-result behavior.

Reviewer: Above we say
>- Only packages with one or more sponsorship urls are included. 

Can this all be combined into a single note?

> #### Testing strategy

Reviewer: Let's note that the envvar can be configured to DEV or similar as well for aiding in testing on our own.
For automated testing, we should be able to mock registration and not use any live package source.

- Validate command output for packages with no sponsorship information, a single sponsorship URL, and multiple sponsorship URLs.
- Verify top-level and transitive packages are included by default.
- Verify sponsorship rendering behavior in console output.
- Verify JSON output for single and multiple sponsorship URL scenarios while preserving URL order.
- Validate sponsorship metadata is propagated from the Registration index and consumed by the command runner.

## Drawbacks

> 

Reviewer: Some ideas

1. nuget.org is hard-coded, so other sources are not supported today.
2. If using Package Source Mapping, the feature won't work
3. ...

## Rationale and alternatives

> `dotnet package list --sponsor` keeps sponsorship as a read-only package metadata report alongside `--deprecated`, `--vulnerable`, and `--outdated`. 

Reviewer: I think you're listing this as rationale. Maybe clarifying it like,

```suggestion
The rationale for our approach using `dotnet package list --sponsor` keeps sponsorship as a read-only package metadata report alongside `--deprecated`, `--vulnerable`, and `--outdated`. 
```

> A dedicated `dotnet package sponsor` command was considered, but this experience reports information and does not perform a sponsorship action.

Reviewer: I think the reason here isn't accurate. We consider it a future possibility that can support more interactive UI (like selecting/launching a sponsor hyperlink from the CLI). The reason we chose `package list` was more around consistency with existing options deprecated/vulnerable.

```suggestion
A dedicated `dotnet package sponsor` command was considered, ....
```

A dedicated command can be reconsidered if interactive sponsorship workflows are added later.

## Prior Art

- [2023 community sponsor-link proposal](https://github.com/NuGet/Home/blob/sponsor-link/proposed/2023/sponsor-link.md)
- [`npm fund`](https://docs.npmjs.com/cli/v10/commands/npm-fund/)
- Existing report-style CLI surfaces: `--deprecated`, `--vulnerable`, `--outdated`

## Unresolved Questions

> If future versions support multiple sponsorship sources, how should results be merged, and how should “no sponsorship URLs” be distinguished from “source does not support sponsorship”?

Reviewer: > and how should “no sponsorship URLs” be distinguished from “source does not support sponsorship”?

We have listed 2 distinct error messages already in this spec.

    }
  ],
  "projects": [
    {
      "path": "/path/to/Contoso.App.csproj",
      "sources": [
        {
          "source": "https://api.example.org/v3/index.json",
          "sponsorablePackages": [
            {
              "id": "Contoso.Tools",
              "urls": [
                "https://github.com/sponsors/contoso"
              ]
            },
            {
              "id": "Contoso.Utility",
              "urls": [
                "https://github.com/sponsors/contoso"
              ]
            }
          ]
        },
        {
          "source": "https://www.myget.org/F/contoso/api/v3/index.json",
          "sponsorablePackages": [
            {
              "id": "Contoso.Utility",
              "urls": [
                "https://buymeacoffee.com/contoso"
              ]
            }
          ]
        },
        {
          "source": "https://packages.contoso.com/v3/index.json",
          "sponsorablePackages": []
        }
      ]
    }
  ]
}
```

**Package Source Mapping and using `--source`**

Registration requests may disclose package IDs to package sources; therefore, sponsorship reporting will honor Package Source Mapping (PSM). 

> When PSM is disabled, each package is queried against each selected source. 

Reviewer: ```suggestion
When PSM is disabled, each package is queried against all configured package sources. 
```

> When PSM is enabled, the client queries only sources that are selected for the command and mapped to that package. 

Reviewer: ```suggestion
When PSM is enabled, the client queries only sources that are mapped to that package. 
```

> Sources that are specified with `--source` replace the configured source selection, but they will not bypass PSM. 

Reviewer: I recall there was disagreement, but PSM being enabled should just be an immediate error with `--source` in my mind.

It's a very ambiguous scenario to say "don't use my package sources but do use my source mappings".

However, I don't think we should _ignore_ the source mappings ever. This is why I think we should just "error" and say you've specified a source which isn't part of the configuration.

> If there is a package not mapped to any selected source, the package is not queried and the console will output a warning: 

Reviewer: PSM by design requires mappings for all packages in order to restore.

```text

> Package `Contoso.Private` was not queried for sponsorship details because none of the selected package sources match its Package Source Mapping.

Reviewer: These seems a bit noisy. Maybe just 1 message saying that Package Source Mapping is applied at the start of the command?

- Load PSM from the nuget settings used by the command.
  - For each package ID, `ListPackageCommandRunner` queries only the selected sources whose configured source name matches that package's source mappings. 
- Add `RegistrationIndexMetadata` for the optional root `metadata` object. 
  - The first property is the ordered `sponsorshipUrls` string collection.
  - Add optional `metadata` property to `RegistrationIndex`.
- Expose root index metadata through a new `RegistrationResourceV3`.
  - Future package-level properties added to the root `metadata` object can use the same abstraction.
- Add sponsorship-specific query path in `ListPackageCommandRunner`.
  - Query each `(source, package id)` pair once and retain the source with the ordered sponsorship URLs. 
- Extend `ListPackageProjectModel` to retain each source and its query results. 
  - Extend `ListReportPackage` with grouped sponsorship details.
- Render console sponsorship output through `ListPackageConsoleRenderer` and `ProjectPackagesPrintUtility`.
- Render JSON sponsorship output through `ListPackageJsonRenderer`.

#### Data assumptions

> - Proposed contract is an optional array of strings on the packages Registration index root exposed as an ordered `IReadOnlyList<string>`.

Reviewer: nit

```suggestion
- Proposed contract is an optional array of strings on the package's Registration index root exposed as an ordered `IReadOnlyList<string>`.
```

A source that does not expose Registration or has authentication, network, or protocol failures uses the same list-package failure behavior. 

#### Testing strategy

- Validate output for packages with no sponsorship information, a single sponsorship URL, and multiple sponsorship URLs in the CLI and JSON.
- Verify top-level and transitive packages are included by default.
- Cover missing or null `metadata` and `sponsorshipUrls` behavior, along with empty arrays or malformed JSON.
- Verify enabled configured source selection including packages mapped to one source, multiple sources, and sources excluded by `--source`.
- Validate sponsorship metadata is propagated from the Registration index and consumed by the command runner.

## Drawbacks

- At this time, nuget.org is the only source that supports sponsorship details in the Registration index root. 
  - Other sources will produce empty results until they implement the same properties. 
  - Sponsorship details from an upstream source are not shown unless that source exposes those details through its own Registration response. 

> - PSM and an explicit `--source` can intentionally exclude sources that may contain sponsorship details.

Reviewer: This is the design of PSM, so I think the drawback is just if a customer saves a nuget.org package to a private feed, then creates a source mapping to the private feed, the sponsor links will never be shown.

# Implementing NuGet.Client CLI Sponsorship Support

- Author: [@kalebfik](https://github.com/kalebfik)
- GitHub Issue: [NuGet/NuGetGallery#10703](https://github.com/NuGet/NuGetGallery/issues/10703)

> - [Accompanying Functional Spec](https://github.com/NuGet/Home/blob/dev-kalebfika-sponsorCliSpec/accepted/2026/sponsorship-cli.md)

Reviewer: nit: Your branch will get deleted post-merge

```suggestion
- [Functional Spec](https://github.com/NuGet/Home/blob/dev/accepted/2026/sponsorship-cli.md)
```

## Summary

Add sponsorship reporting to the NuGet CLI for installed packages.

The command surface will be presented as `dotnet package list --sponsor`.

It will reuse the same metadata-fetch and rendering pipeline, with:

- A `Sponsor` column in console output.
- A `sponsorablePackages` array in JSON output.

## Explanation

### Functional Explanation

When a user runs `dotnet package list --sponsor`, the CLI will examine the packages in the project or solution and retrieve sponsorship links from the selected package sources.
The results are then grouped by the source that provided the sponsorship details.
For packages returned from a source, URLs preserve the order returned by that source.

The command selects enabled package sources from NuGet configuration.
When a source is specified using `--source <SOURCE>`, only that source is queried for sponsorship information.

Console output includes only sources that return one or more sponsorship URLs.
If none of the selected sources return sponsorship details for a project, the CLI displays:

```text
// sample response
No sponsorship details were returned using the following package sources:
  https://api.example.org/v3/index.json

> Consider specifying an additional package source that provides sponsorship metadata, for example: `--source https://api.nuget.org/v3/index.json`.

Reviewer: Just to clarify this one, we should only include this line if `https://api.nuget.org/v3/index.json` is not already a configured source.

In other words, we don't want to say, no sponsorship details found for nuget.org, then immediately suggest including nuget.org.
